### PR TITLE
[Layer dependencies] Emit QgsVectorLayer::dataChanged on commitChanges

### LINKED ISSUE #57784: Rubberband/snapping using old nodes if a geometry is modified outside QGIS
### What is the bug or the crash?

If a geometry is modified in a database after saving layer edits in QGIS the geometry isn't updated in edit/snapping layers.

editbug

In the example case below, there is a PostGIS table with a trigger that makes a translation for each row inserted or updated on the table. The polygon itself is updated in QGIS correctly but the edit/snapping layer has still the old geometry. This causes most probably for example the issue described in #57181.

### Steps to reproduce the issue

1. Create a table with a trigger translating the geometry in each insert or update.

```
-- create a trigger function to translate polygons
CREATE OR REPLACE FUNCTION public.translate_geom()
RETURNS trigger AS $$
BEGIN
  NEW.geom = ST_Translate(NEW.geom, 1000, 0);
  RETURN NEW;
END; $$ LANGUAGE 'plpgsql';

-- create a test table
CREATE TABLE translated_polygons
(
    id bigserial NOT NULL,
    geom geometry(POLYGON, 3067),
    CONSTRAINT "PK_translated_polygons" PRIMARY KEY (id)
);

-- create a trigger
CREATE TRIGGER translate_geom
BEFORE INSERT OR UPDATE ON translated_polygons
FOR EACH ROW EXECUTE PROCEDURE public.translate_geom();
```
2. Open the table in QGIS.

3. Start editing.

4. Digitize a polygon.

5. Save layer edits.

6. Move one vertex (or modify existing geometry some other way).

7. Save layer edits.

8. The polygon is shown correctly but for example the vertex tool and snapping use old coordinates.

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.36.3-Maidenhead | QGIS code revision | 2df9655469b
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.10.12
Compiled against GDAL/OGR | 3.4.1 | Running against GDAL/OGR | 3.4.3
PROJ version | 8.2.1
EPSG Registry database version | v10.076 (2022-08-31)
Compiled against GEOS | 3.10.2-CAPI-1.16.0 | Running against GEOS | 3.11.1-CAPI-1.17.1
SQLite version | 3.37.2
Compiled against PDAL | 2.3.0 | Running against PDAL | 2.5.0
PostgreSQL client version | 14.11 (Ubuntu 14.11-0ubuntu0.22.04.1)
SpatiaLite version | 5.0.1
QWT version | 6.1.4
QScintilla2 version | 2.11.6
OS version | Ubuntu 22.04.4 LTS
  |   |   |  
Active Python plugins
processing | 2.12.99
grassprovider | 2.12.99
db_manager | 0.1.20
MetaSearch | 0.3.6

</body></html>QGIS version
3.36.3-Maidenhead
QGIS code revision
[2df9655469b](https://github.com/qgis/QGIS/commit/2df9655469b)
Qt version
5.15.3
Python version
3.10.12
Compiled against GDAL/OGR
3.4.1
Running against GDAL/OGR
3.4.3
PROJ version
8.2.1
EPSG Registry database version
v10.076 (2022-08-31)
Compiled against GEOS
3.10.2-CAPI-1.16.0
Running against GEOS
3.11.1-CAPI-1.17.1
SQLite version
3.37.2
Compiled against PDAL
2.3.0
Running against PDAL
2.5.0
PostgreSQL client version
14.11 (Ubuntu 14.11-0ubuntu0.22.04.1)
SpatiaLite version
5.0.1
QWT version
6.1.4
QScintilla2 version
2.11.6
OS version
Ubuntu 22.04.4 LTS

Active Python plugins
processing
2.12.99
grassprovider
2.12.99
db_manager
0.1.20
MetaSearch
0.3.6

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

It seems that the bug is caused by QgsPointLocator that doesn't get the modified geometry. When a new feature is added the corresponding geometry is added to QgsPointLocator in the function onFeatureAdded().

https://github.com/qgis/QGIS/blob/29bb1944f02704ebad49ad0199fcf7b681bd55fe/src/core/qgspointlocator.cpp#L1261-L1277

When modifying an existing geometry this isn't handled. I'm not sure about the most feasible solution. One solution could be emitting a signal for the changed geometry in QgsVectorLayerEditBuffer::commitChangesChangeAttributes() but still we would need to ensure that QgsPointLocator gets the modified geometry.

### PR DESCRIPTION
**Pull request by Djedouas**
## Description

Fixes #57784 where even if the layer is checked as a dependency layer, there is no reloading of data when the layer is saved.

With this change, the dataChanged signal is emitted on commitChanges, so that when a feature is externally modified on save (database trigger for example) we have index and cache reconstruction.

---

Related PRs:
- https://github.com/qgis/QGIS/pull/37475
- https://github.com/qgis/QGIS/pull/41463

This change is kind of following this particular comment/discussion https://github.com/qgis/QGIS/pull/41463#issuecomment-777094703 

See the tests explaining the multiple emissions of `dataChanged`.

Ping @suricactus @m-kuhn @MorriganR @troopa81 @3nids for review

### CODE REVIEW

File: tests/src/python/test_layer_dependencies.py [hunk 1]
```diff
@@ -26,7 +26,6 @@
     QgsRectangle,
     QgsSnappingConfig,
     QgsSnappingUtils,
-    QgsTolerance,
     QgsVectorLayer,
```
Comment (Djedouas) [inline tests/src/python/test_layer_dependencies.py:29]: unused import

**Comment by troopa81** [comment]
@Djedouas After a second thought on our discussion I'm wondering if we could not fire only one dataChange on commit instead of 2 when adding (delete the -1 feature +add the inserted one) and 0 (3 with your proposal) when commit. Of course, without changing the general behavior when it's not about committing.

We could use the `mCommitChangesActive` boolean to have a different behavior in emitDataChanged() method. This behavior would not reset `mDataChangedFired` to false and it would be done when `mCommitChangesActive` is reset after commit ? Would that work?

**Comment by nyalldawson** [comment]
This is potentially a big speed regression -- eg QgsPointLocator connects to QgsVectorLayer::dataChanged and destroys the index when it's emitted, and the attribute table (via QgsVectorLayerCache) will also trigger a full rebuild. My understanding was that dataChanged should only be used when we don't know what's changed, but when we DO know what's changed the atomic signals (featureAdded, etc) are better because we can handle them in optimised ways.

Isn't all this supposed to be handled by the logic in QgsVectorLayer::setDependencies anyway? That's setting up manual connections to emit dataChanged when the atomic signals are received from a dependent layer.

**Comment by troopa81** [comment]
> My understanding was that dataChanged should only be used when we don't know what's changed, but when we DO know what's changed the atomic signals (featureAdded, etc) are better because we can handle them in optimised ways.

That's the case here, a dependency has been committed, and so eventually the layer depending on it has also changed (because of trigger for instance) and need to fire dataChanged . This PR doesn't change this.

> Isn't all this supposed to be handled by the logic in QgsVectorLayer::setDependencies anyway? That's setting up manual connections to emit dataChanged when the atomic signals are received from a dependent layer.

Yes, that's what it's intended here (don't know why we reloadData() and not just fire dataChanged signal in that case). I don't see any regression here, except IMHO we fire too many dataChanged per event.

**Comment by nyalldawson** [comment]
I'm confused now -- was a force push used here to change the approach? The current changes don't seem to match with the PR description (or more recollections from previously reviewing this)

**Comment by Djedouas** [comment]
Nothing changed since the PR creation

File: src/core/vector/qgsvectorlayer.cpp [hunk 2]
```diff
@@ -6109,22 +6109,21 @@ void QgsVectorLayer::emitDataChanged()
   if ( mDataChangedFired )
     return;
 
+  // If we are asked to fire dataChanged from a layer we depend on,
+  // be sure that this layer is not in the process of committing its changes, because
+  // we will be asked to fire dataChanged at the end of his commit, and we don't
+  // want to fire this signal more than necessary.
+  if ( QgsVectorLayer *layerWeDependUpon = dynamic_cast<QgsVectorLayer *>( sender() ) )
+    if ( layerWeDependUpon->mCommitChangesActive )
```
Comment (troopa81) [inline src/core/vector/qgsvectorlayer.cpp:6117]:
```suggestion
  if ( ( QgsVectorLayer *layerWeDependUpon = qobject_cast<QgsVectorLayer *>( sender() ) ) && layerWeDependUpon->mCommitChangesActive )
```

one lining and qobject_cast is supposedly better than dynamic_cast when it comes to QObject

**Review by troopa81 — COMMENTED** [review]
@nyalldawson @m-kuhn 

We discussed this with @Djedouas , I think the proposal is good as it implies only one additional signal firing when committing instead of 3 (feature added+deleted+commit) when it comes to new feature.

**Review by troopa81 — APPROVED** [review]

**Pull request merged**